diff --git a/pkgs/tools/typesetting/tectonic/biber.nix b/pkgs/by-name/bi/biber-for-tectonic/package.nix
similarity index 100%
rename from pkgs/tools/typesetting/tectonic/biber.nix
rename to pkgs/by-name/bi/biber-for-tectonic/package.nix
diff --git a/pkgs/tools/typesetting/tectonic/default.nix b/pkgs/by-name/te/tectonic-unwrapped/package.nix
similarity index 88%
rename from pkgs/tools/typesetting/tectonic/default.nix
rename to pkgs/by-name/te/tectonic-unwrapped/package.nix
index 58c92c1dc4afef2f2286d3db63374f61f44a5df0..e9b9dcb2c81579314a1ad8f1453a551d168ebf5f 100644
--- a/pkgs/tools/typesetting/tectonic/default.nix
+++ b/pkgs/by-name/te/tectonic-unwrapped/package.nix
@@ -1,8 +1,8 @@
 /*
   This file provides the `tectonic-unwrapped` package. On the other hand,
-  the `tectonic` package is defined in `./wrapper.nix`, by wrapping
-  - [`tectonic-unwrapped`](./default.nix) i.e. this package, and
-  - [`biber-for-tectonic`](./biber.nix),
+  the `tectonic` package is defined in `../tectonic/package.nix`, by wrapping
+  - [`tectonic-unwrapped`](./package.nix) i.e. this package, and
+  - [`biber-for-tectonic`](../../bi/biber-for-tectonic/package.nix),
     which provides a compatible version of `biber`.
 */
 
@@ -13,7 +13,7 @@
   rustPlatform,
   darwin,
   fontconfig,
-  harfbuzz,
+  harfbuzzFull,
   openssl,
   pkg-config,
   icu,
@@ -60,7 +60,7 @@ rustPlatform.buildRustPackage rec {
     [
       icu
       fontconfig
-      harfbuzz
+      harfbuzzFull
       openssl
     ]
     ++ lib.optionals stdenv.hostPlatform.isDarwin (
@@ -86,13 +86,13 @@ rustPlatform.buildRustPackage rec {
 
   doCheck = true;
 
-  meta = with lib; {
+  meta = {
     description = "Modernized, complete, self-contained TeX/LaTeX engine, powered by XeTeX and TeXLive";
     homepage = "https://tectonic-typesetting.github.io/";
     changelog = "https://github.com/tectonic-typesetting/tectonic/blob/tectonic@${version}/CHANGELOG.md";
-    license = with licenses; [ mit ];
+    license = with lib.licenses; [ mit ];
     mainProgram = "tectonic";
-    maintainers = with maintainers; [
+    maintainers = with lib.maintainers; [
       lluchs
       doronbehar
       bryango
diff --git a/pkgs/tools/typesetting/tectonic/wrapper.nix b/pkgs/by-name/te/tectonic/package.nix
similarity index 100%
rename from pkgs/tools/typesetting/tectonic/wrapper.nix
rename to pkgs/by-name/te/tectonic/package.nix
diff --git a/pkgs/tools/typesetting/tectonic/tests.nix b/pkgs/by-name/te/tectonic/tests.nix
similarity index 100%
rename from pkgs/tools/typesetting/tectonic/tests.nix
rename to pkgs/by-name/te/tectonic/tests.nix
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 16ebad70b3ffc1351d4e4c6a704d9d9a6598642b..51f20ece9e3a76142fd66b7847a94fa77954705a 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -2677,8 +2677,6 @@ with pkgs;
   beetsPackages = lib.recurseIntoAttrs (callPackage ../tools/audio/beets { });
   inherit (beetsPackages) beets beets-unstable;
 
-  biber-for-tectonic = callPackage ../tools/typesetting/tectonic/biber.nix { };
-
   binlore = callPackage ../development/tools/analysis/binlore { };
 
   birdfont = callPackage ../tools/misc/birdfont { };
@@ -10683,12 +10681,6 @@ with pkgs;
 
   tclap_1_4 = callPackage ../development/libraries/tclap/1.4.nix { };
 
-  tectonic = callPackage ../tools/typesetting/tectonic/wrapper.nix { };
-
-  tectonic-unwrapped = callPackage ../tools/typesetting/tectonic {
-    harfbuzz = harfbuzzFull;
-  };
-
   termbench-pro = callPackage ../by-name/te/termbench-pro/package.nix {
     stdenv = if stdenv.hostPlatform.isDarwin then llvmPackages_17.stdenv else stdenv;
     fmt = fmt_11;
